File: server/src/andes_app/core/jobs.py
"""Per-session in-memory job registry — Unit 1 of v3.1 UX overhaul.

The registry tracks every routine invocation (PF, EIG, CPF, SE, snapshots,
bundles, clone edits, etc.) with a stable ``job_id``, lifecycle status,
started/ended timestamps, and an optional ``ProblemDetails`` dict on failure.

Retention is *semantic*, not pure-FIFO (KTD-19 of the v3.1 plan; sticky-first
failures motivated by the adversarial review):

- Never evict ``pending`` or ``running`` jobs.
- ``mark_failed`` coalesces duplicates by ``(kind, category, detail)``
  signature: a fresh failure matching an existing record collapses into
  that record (the first-occurrence ``started_at`` and problem dict are
  preserved; ``repeated_count`` increments). This defends against cascade
  context loss (failure #1 holds the diagnostic; #21 is the symptom — FIFO
  would silently drop the diagnostic).
- Among successful jobs, FIFO eviction at ``MAX_SUCCESSFUL`` keeps the
  most recent.
- Combined ring buffer caps at ``MAX_TOTAL``; on overflow, oldest terminal
  records evict first, in-flight records are never touched.

The registry is thread-safe (a single ``threading.Lock`` guards every
mutation). All reads return defensive copies so callers can't mutate
internal state through a returned reference.

This module is read-only with respect to the rest of the substrate: no
worker, no FastAPI, no ANDES. Unit 5a (Phase 2) adds the ``/jobs`` routes
that expose the registry over HTTP + WS.
"""

from __future__ import annotations

import threading
import time
import uuid
from copy import deepcopy
from dataclasses import dataclass, field
from typing import Any, Literal

JobKind = Literal[
    # routines
    "pflow",
    "tds-batch",
    "tds-stream",
    "eig",
    "cpf",
    "cpf-qv",
    "se",
    "se-measurements",
    "sweep",
    # state ops
    "snapshot-save",
    "snapshot-restore",
    "snapshot-delete",
    "bundle-export",
    "bundle-import",
    "case-load",
    "case-reload",
    "case-save",
    # edits + addfile
    "element-add",
    "element-edit",
    "element-delete",
    "element-undo",
    "disturbance-commit",
    "pmu-add",
    "pmu-delete",
    "profile-upload",
    "profile-add",
    "profile-delete",
    # clone-on-write (Phase 6)
    "clone-init",
    "clone-edit",
    "clone-undo",
    "clone-redo",
    "clone-save-as",
    "clone-reset",
]


JobStatus = Literal["pending", "running", "done", "failed", "cancelled"]


# Retention knobs (KTD-19). Distinct-failure cap defends against cascade
# context loss; successful FIFO is the "rolling history" the activity panel
# scrolls through; total cap is the safety net.
MAX_FAILED_DISTINCT: int = 20
MAX_SUCCESSFUL: int = 50
MAX_TOTAL: int = 100


@dataclass
class JobRecord:
    """One job's lifecycle row.

    Returned via ``_JobRegistry.get_job`` / ``list_jobs`` as a defensive
    copy; mutating a returned record does not affect the registry.
    """

    id: str
    kind: JobKind
    status: JobStatus
    started_at: float
    updated_at: float
    can_cancel: bool
    request_summary: dict[str, Any] = field(default_factory=dict)
    progress: float | None = None
    ended_at: float | None = None
    result_ref: str | None = None
    problem: dict[str, Any] | None = None
    # Sticky-first coalescing counter. A fresh failure whose signature
    # matches this record bumps the count rather than displacing the
    # first-occurrence diagnostic.
    repeated_count: int = 0
    # Originating session id. Stamped on records that live in the manager-wide
    # global registry (session-mutating jobs, KTD-20) so the per-session HTTP
    # surface can filter the shared registry to the owning session — otherwise
    # every session's ``GET /jobs`` would leak every other session's global
    # jobs. ``None`` for per-session-registry records (no filtering needed
    # there — the registry is already session-scoped). Internal: NOT serialized
    # onto the wire ``JobRecordSchema`` (which only declares its own fields).
    origin_session_id: str | None = None


def _failure_signature(record: JobRecord) -> tuple[str, str, str] | None:
    """Stable signature for a failed job's error.

    Two records with the same ``(kind, category, detail)`` are treated
    as repeats of the same underlying problem. Returns ``None`` for
    records that aren't ``failed``.
    """
    if record.status != "failed" or record.problem is None:
        return None
    category = str(record.problem.get("category") or "")
    detail = str(record.problem.get("detail") or "")
    return (record.kind, category, detail)


class _JobRegistry:
    """Per-session in-memory job log."""

    def __init__(self) -> None:
        self._records: dict[str, JobRecord] = {}
        self._lock = threading.Lock()

    def register_job(
        self,
        *,
        kind: JobKind,
        can_cancel: bool,
        request_summary: dict[str, Any] | None = None,
        job_id: str | None = None,
        origin_session_id: str | None = None,
    ) -> str:
        """Register a fresh ``pending`` job. Returns the new ``job_id``.

        ``job_id`` defaults to a freshly-minted uuid. v3.1 Unit 5c passes an
        explicit id so the streaming-TDS / sweep jobs can alias the registry
        ``job_id`` onto the pre-existing ``run_id`` / ``sweep_id`` (same value
        across both fields). Re-registering an already-present id is a no-op
        that returns the existing id rather than clobbering its lifecycle.

        ``origin_session_id`` stamps the owning session onto the record. It is
        passed for records that land in the manager-wide global registry
        (session-mutating jobs) so the per-session HTTP surface can filter the
        shared registry to its own session.
        """
        now = time.monotonic()
        if job_id is None:
            job_id = str(uuid.uuid4())
        with self._lock:
            if job_id in self._records:
                return job_id
            self._records[job_id] = JobRecord(
                id=job_id,
                kind=kind,
                status="pending",
                started_at=now,
                updated_at=now,
                can_cancel=can_cancel,
                request_summary=dict(request_summary or {}),
                origin_session_id=origin_session_id,
            )
            self._evict_if_over_cap()
        return job_id

    def mark_running(self, job_id: str) -> None:
        with self._lock:
            record = self._records.get(job_id)
            if record is None or record.status != "pending":
                return
            record.status = "running"
            record.updated_at = time.monotonic()

    def mark_done(self, job_id: str, *, result_ref: str | None = None) -> None:
        with self._lock:
            record = self._records.get(job_id)
            if record is None or record.status in ("done", "failed", "cancelled"):
                return
            now = time.monotonic()
            record.status = "done"
            record.result_ref = result_ref
            record.updated_at = now
            record.ended_at = now
            self._evict_if_over_cap()

    def mark_failed(self, job_id: str, *, problem: dict[str, Any]) -> str:
        """Mark the job failed and apply sticky-first signature coalescing.

        Returns the ``job_id`` of the SURVIVING record — equal to the passed
        ``job_id`` normally, or (on coalescing) the id of the PRIOR record the
        failure collapsed into. Callers MUST broadcast the RETURNED id (not the
        passed one) so the live ``/jobs/events`` feed always reflects the
        transition: when the passed id is coalesced away the surviving prior
        record (with its bumped ``repeated_count``) is the authoritative one,
        and re-reading the deleted passed id would yield ``None`` → a dropped
        terminal envelope leaving WS-only panels spinning forever.

        If an existing failed record shares this record's
        ``(kind, category, detail)`` signature, the new record is
        discarded and the prior record's ``repeated_count`` is
        incremented (preserving the original ``started_at`` and problem
        dict so the diagnostic isn't lost).

        A no-op (returns the passed ``job_id``) when the record is already
        terminal — guarding against a post-cancel driver error flipping a
        ``cancelled`` / ``done`` record back to ``failed`` (the cancel path
        does NOT abort the underlying work synchronously, so a late driver
        error can race the cancel transition).
        """
        with self._lock:
            record = self._records.get(job_id)
            if record is None:
                return job_id
            if record.status in ("done", "failed", "cancelled"):
                # Already terminal: do not overwrite. Symmetric with
                # ``mark_done`` / ``mark_cancelled``; closes the race where a
                # late driver error flips a ``cancelled`` record to ``failed``.
                return job_id
            now = time.monotonic()
            record.problem = dict(problem)
            record.status = "failed"
            record.updated_at = now
            record.ended_at = now
            signature = _failure_signature(record)
            survivor_id = job_id
            if signature is not None:
                for prior_id, prior in list(self._records.items()):
                    if prior_id == job_id:
                        continue
                    if _failure_signature(prior) == signature:
                        prior.repeated_count += 1
                        prior.updated_at = now
                        del self._records[job_id]
                        # The prior record is the survivor; callers broadcast
                        # it so the coalesced failure still reaches subscribers.
                        survivor_id = prior_id
                        break
            self._evict_if_over_cap()
        return survivor_id

    def mark_cancelled(self, job_id: str) -> None:
        with self._lock:
            record = self._records.get(job_id)
            if record is None or record.status in ("done", "failed", "cancelled"):
                return
            now = time.monotonic()
            record.status = "cancelled"
            record.updated_at = now
            record.ended_at = now
            self._evict_if_over_cap()

    def update_progress(self, job_id: str, progress: float) -> None:
        """Update an in-flight job's progress. Clamps to ``[0.0, 1.0]``;
        no-op when the job is terminal."""
        clamped = max(0.0, min(1.0, progress))
        with self._lock:
            record = self._records.get(job_id)
            if record is None or record.status not in ("pending", "running"):
                return
            record.progress = clamped
            record.updated_at = time.monotonic()

    def get_job(self, job_id: str) -> JobRecord | None:
        with self._lock:
            record = self._records.get(job_id)
            if record is None:
                return None
            return _copy_record(record)

    def list_jobs(
        self,
        *,
        kind: JobKind | None = None,
        status: JobStatus | None = None,
    ) -> list[JobRecord]:
        """Return jobs matching the optional filters, in insertion order."""
        with self._lock:
            out: list[JobRecord] = []
            for record in self._records.values():
                if kind is not None and record.kind != kind:
                    continue
                if status is not None and record.status != status:
                    continue
                out.append(_copy_record(record))
            return out

    def _evict_if_over_cap(self) -> None:
        """Apply KTD-19 retention. Must be called with ``self._lock`` held.

        Sequence (each step preserves the never-evict-in-flight invariant):

        1. Among ``done`` records, evict oldest until ``MAX_SUCCESSFUL``.
        2. Among ``failed`` records, evict oldest until ``MAX_FAILED_DISTINCT``.
           (Sticky-first coalescing in ``mark_failed`` already collapses
           duplicates by signature, so this only fires when ≥21 distinct
           error signatures have accumulated.)
        3. If the overall record count still exceeds ``MAX_TOTAL``, evict
           the oldest terminal records (any bucket) until at cap.
        """
        done_ids = [k for k, r in self._records.items() if r.status == "done"]
        if len(done_ids) > MAX_SUCCESSFUL:
            for stale in done_ids[: len(done_ids) - MAX_SUCCESSFUL]:
                del self._records[stale]

        failed_ids = [k for k, r in self._records.items() if r.status == "failed"]
        if len(failed_ids) > MAX_FAILED_DISTINCT:
            for stale in failed_ids[: len(failed_ids) - MAX_FAILED_DISTINCT]:
                del self._records[stale]

        if len(self._records) > MAX_TOTAL:
            terminal_ids = [
                k
                for k, r in self._records.items()
                if r.status in ("done", "failed", "cancelled")
            ]
            overflow = len(self._records) - MAX_TOTAL
            for stale in terminal_ids[:overflow]:
                del self._records[stale]


def _copy_record(record: JobRecord) -> JobRecord:
    return JobRecord(
        id=record.id,
        kind=record.kind,
        status=record.status,
        started_at=record.started_at,
        updated_at=record.updated_at,
        can_cancel=record.can_cancel,
        request_summary=deepcopy(record.request_summary),
        progress=record.progress,
        ended_at=record.ended_at,
        result_ref=record.result_ref,
        problem=deepcopy(record.problem) if record.problem else None,
        repeated_count=record.repeated_count,
        origin_session_id=record.origin_session_id,
    )

File: web/src/store/jobs.ts
/**
 * Jobs slice (v3.1 Phase 3, Unit 6).
 *
 * Aggregates + indexes every routine / streaming / sweep / state-op / edit
 * invocation under a single ``JobRecord`` map keyed by ``job_id``, mirroring
 * the substrate ``_JobRegistry`` (``server/src/andes_app/core/jobs.py``). This
 * slice is the canonical client-side view that powers the Activity panel
 * (Unit 11) and lets ``SessionBusy`` be tracked as a job *state* rather than
 * surfacing as unhandled boot-409 console noise.
 *
 * Two write paths converge here and reconcile by ``job_id``:
 *
 * 1. **Mutation hooks** (``api/queries.ts``) register a placeholder
 *    ``JobRecord`` on ``onMutate`` for instant optimistic feedback (the
 *    server ``job_id`` may not be known yet, so a temp id is minted), then
 *    upgrade to the substrate ``job_id`` once the response lands.
 * 2. **``JobStream``** events (``streaming/JobStream.ts``) carry the
 *    canonical substrate state and overwrite/merge by ``job_id``.
 *
 * Whichever path arrives first, the other merges onto it (``upsertJob`` /
 * ``reconcileJob``). This slice does NOT replace ``useRunsStore`` (TDS
 * frame storage) or ``useSweepStore`` (per-iteration sweep state) — it
 * aggregates and indexes them by job id.
 *
 * **SECURITY (F2 / KTD-16):** the full ``JobRecord`` map carries
 * ``request_summary`` payloads, ``problem`` error details, and
 * ``result_ref`` results. These are IN-MEMORY ONLY and MUST NOT be
 * persisted to localStorage. The ``partialize`` whitelist below is
 * strictly limited to display-state metadata (``dismissedJobIds``) — the
 * record map is never serialized.
 *
 * Retention mirrors the substrate's KTD-19 cap (oldest *terminal* first;
 * in-flight records are NEVER evicted) at ``MAX_TERMINAL`` entries.
 */
import { create } from 'zustand';
import { persist, createJSONStorage } from 'zustand/middleware';

/**
 * Routine / op discriminator. Mirrors ``andes_app.core.jobs.JobKind`` 1:1
 * so a substrate event's ``kind`` round-trips without translation. Kept as
 * a string-literal union (forward-compat: an unknown future kind from the
 * wire still stores; the union narrows known kinds for exhaustive UI
 * switches).
 */
export type JobKind =
  // routines
  | 'pflow'
  | 'tds-batch'
  | 'tds-stream'
  | 'eig'
  | 'cpf'
  | 'cpf-qv'
  | 'se'
  | 'se-measurements'
  | 'sweep'
  // state ops
  | 'snapshot-save'
  | 'snapshot-restore'
  | 'snapshot-delete'
  | 'bundle-export'
  | 'bundle-import'
  | 'case-load'
  | 'case-reload'
  | 'case-save'
  // edits + addfile
  | 'element-add'
  | 'element-edit'
  | 'element-delete'
  | 'element-undo'
  | 'disturbance-commit'
  | 'pmu-add'
  | 'pmu-delete'
  | 'profile-upload'
  | 'profile-add'
  | 'profile-delete'
  // clone-on-write (Phase 6)
  | 'clone-init'
  | 'clone-edit'
  | 'clone-undo'
  | 'clone-redo'
  | 'clone-save-as'
  | 'clone-reset';

/** Lifecycle state. Mirrors ``andes_app.core.jobs.JobStatus``. */
export type JobStatus = 'pending' | 'running' | 'done' | 'failed' | 'cancelled';

/**
 * The set of terminal statuses. A record in any of these states is a
 * candidate for retention eviction; in-flight (``pending`` / ``running``)
 * records are NEVER evicted.
 */
export const TERMINAL_STATUSES: ReadonlySet<JobStatus> = new Set<JobStatus>([
  'done',
  'failed',
  'cancelled',
]);

export function isTerminalStatus(status: JobStatus): boolean {
  return TERMINAL_STATUSES.has(status);
}

/**
 * Minimal ProblemDetails shape carried on a failed job. Mirrors the
 * substrate's ``ProblemDetails`` envelope loosely — the Activity panel's
 * ``<ProblemDetailsErrorSurface>`` (Unit 7) reads ``title`` / ``detail`` /
 * ``recovery`` off this. Kept structurally open (``[extra: string]``) so a
 * forward-compat field doesn't drop on the wire.
 */
export interface JobProblem {
  type?: string;
  title?: string;
  status?: number;
  detail?: string | null;
  instance?: string | null;
  category?: string;
  recovery?: unknown;
  [extra: string]: unknown;
}

/**
 * Client-side ``JobRecord`` mirroring the substrate's
 * ``JobRecordSchema`` wire shape. Field names match the wire snake_case so
 * a substrate event/list payload maps on without per-field translation.
 *
 * ``request_summary`` / ``problem`` / ``result_ref`` are in-memory ONLY
 * (security F2): they MUST NOT appear in the persist whitelist.
 */
export interface JobRecord {
  /** Canonical substrate ``job_id``, OR a client temp id (``local:<uuid>``)
   *  until the server id arrives via the mutation response. */
  id: string;
  kind: JobKind;
  status: JobStatus;
  /** Monotonic-ish seconds. For client placeholders we stamp ``Date.now()/1000``
   *  so chronological ordering against substrate timestamps stays sane
   *  enough for display; the wire ``started_at`` overwrites on reconcile. */
  started_at: number;
  updated_at: number;
  ended_at?: number;
  can_cancel: boolean;
  progress?: number;
  request_summary: Record<string, unknown>;
  result_ref?: string;
  problem?: JobProblem | null;
  repeated_count: number;
  /**
   * True while this record is a client-minted placeholder whose canonical
   * ``job_id`` has not yet arrived. Cleared on ``reconcileJob`` when the
   * substrate id replaces the temp id. Display-only — never persisted.
   */
  isPlaceholder?: boolean;
}

/**
 * Cap on retained TERMINAL records. Mirrors the substrate's ``MAX_TOTAL``
 * ring-buffer cap (``andes_app.core.jobs.MAX_TOTAL``). In-flight records are
 * never counted against this cap and never evicted.
 */
export const MAX_TERMINAL = 100;

/** Prefix for client-minted placeholder ids (pre-server-job_id). */
export const LOCAL_ID_PREFIX = 'local:';

/** Mint a fresh client-side placeholder job id. */
export function mintLocalJobId(): string {
  const rand =
    typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function'
      ? crypto.randomUUID()
      : Math.random().toString(36).slice(2);
  return `${LOCAL_ID_PREFIX}${rand}`;
}

/** Fields a mutation hook supplies when registering an optimistic placeholder. */
export interface AddJobInput {
  /** Caller-supplied id. Defaults to a freshly-minted ``local:`` placeholder. */
  id?: string;
  kind: JobKind;
  /** Defaults to ``pending``. */
  status?: JobStatus;
  can_cancel?: boolean;
  request_summary?: Record<string, unknown>;
  isPlaceholder?: boolean;
  progress?: number;
}

/**
 * A status-event envelope from ``JobStream`` (or a full list-record). The
 * snake_case ``job_id`` mirrors the substrate's ``_job_event_envelope``.
 * Partial-update semantics: ``progress`` / ``problem`` absence means
 * "unchanged" (the substrate only includes them when populated).
 */
export interface JobEventEnvelope {
  job_id: string;
  kind: JobKind;
  status: JobStatus;
  progress?: number;
  problem?: JobProblem | null;
  /** Present on full list payloads (``GET /jobs``), absent on lean events. */
  started_at?: number;
  updated_at?: number;
  ended_at?: number | null;
  can_cancel?: boolean;
  request_summary?: Record<string, unknown>;
  result_ref?: string | null;
  repeated_count?: number;
}

export interface JobsState {
  /** Map ``job_id`` → record. Insertion order = chronological registration. */
  jobs: Record<string, JobRecord>;
  /**
   * Display-state: ids the user has dismissed from the Activity panel's
   * error history. The ONLY field persisted to localStorage (security F2).
   */
  dismissedJobIds: readonly string[];

  /**
   * Register an optimistic placeholder (mutation ``onMutate`` path). Returns
   * the id used (the caller passes it to ``reconcileJob`` once the server
   * id lands). Applies retention after insert.
   */
  addJob: (input: AddJobInput) => string;

  /**
   * Patch an existing record by id. No-op when the id is unknown. Bumps
   * ``updated_at``. Applies retention when the patch moves the record into a
   * terminal state.
   */
  updateJob: (id: string, patch: Partial<JobRecord>) => void;

  /**
   * Upsert a record from a ``JobStream`` event (canonical state). Merges
   * onto an existing record (by ``job_id``) or inserts a fresh one. Applies
   * retention after the write.
   */
  upsertJob: (envelope: JobEventEnvelope) => void;

  /**
   * Reconcile a client placeholder (``tempId``) onto the canonical
   * substrate ``serverId``. If a record under ``serverId`` already exists
   * (the WS event raced ahead), the placeholder's optimistic fields merge
   * onto it and the temp entry is dropped. Otherwise the temp record is
   * re-keyed to ``serverId``. ``patch`` carries any response-derived fields
   * (e.g. ``result_ref``, terminal ``status``).
   */
  reconcileJob: (tempId: string, serverId: string, patch?: Partial<JobRecord>) => void;

  /** Remove a record by id. Idempotent. */
  removeJob: (id: string) => void;

  /** Replace a full snapshot of records (``JobStream`` reconnect re-sync). */
  syncJobs: (envelopes: readonly JobEventEnvelope[]) => void;

  /** Mark a job dismissed (display-state; persisted). Idempotent. */
  dismissJob: (id: string) => void;
  /** Clear a dismissal. Idempotent. */
  undismissJob: (id: string) => void;

  /** Clear every record (auth-clear / session-change cascade). Does NOT clear
   *  ``dismissedJobIds`` — dismissals are display preferences that survive a
   *  session swap; they're cheap and self-healing (a dismissed id that never
   *  recurs is harmless). */
  clearJobs: () => void;
}

// ---- internal helpers -----------------------------------------------------

/**
 * Apply the terminal-retention cap. Returns the next jobs map; mutates
 * nothing on the input. Evicts oldest TERMINAL records first (insertion
 * order = chronological); NEVER evicts in-flight (``pending`` / ``running``)
 * records — mirrors the substrate's ``_evict_if_over_cap`` invariant.
 */
function applyRetention(jobs: Record<string, JobRecord>): Record<string, JobRecord> {
  const terminalIds: string[] = [];
  for (const id of Object.keys(jobs)) {
    if (isTerminalStatus(jobs[id]!.status)) terminalIds.push(id);
  }
  if (terminalIds.length <= MAX_TERMINAL) return jobs;
  const overflow = terminalIds.length - MAX_TERMINAL;
  const next = { ...jobs };
  // Insertion order is chronological for string keys in JS objects, so the
  // leading ``overflow`` terminal ids are the oldest.
  for (let i = 0; i < overflow; i += 1) {
    delete next[terminalIds[i]!];
  }
  return next;
}

/**
 * Build a full ``JobRecord`` from a (possibly-lean) event envelope, merging
 * onto an optional existing record. Absent ``progress`` / ``problem`` /
 * other optional fields are treated as "unchanged" so a lean transition
 * event doesn't wipe data carried by a prior full record.
 */
function mergeEnvelope(envelope: JobEventEnvelope, existing?: JobRecord): JobRecord {
  const now = Date.now() / 1000;
  const base: JobRecord = existing ?? {
    id: envelope.job_id,
    kind: envelope.kind,
    status: envelope.status,
    started_at: envelope.started_at ?? now,
    updated_at: envelope.updated_at ?? now,
    can_cancel: envelope.can_cancel ?? false,
    request_summary: envelope.request_summary ?? {},
    repeated_count: envelope.repeated_count ?? 0,
  };
  const next: JobRecord = {
    ...base,
    id: envelope.job_id,
    kind: envelope.kind,
    status: envelope.status,
    updated_at: envelope.updated_at ?? now,
  };
  // The canonical event is authoritative — a real transition always clears
  // the placeholder flag.
  delete next.isPlaceholder;
  if (envelope.started_at !== undefined) next.started_at = envelope.started_at;
  if (envelope.progress !== undefined) next.progress = envelope.progress;
  if (envelope.can_cancel !== undefined) next.can_cancel = envelope.can_cancel;
  if (envelope.request_summary !== undefined) next.request_summary = envelope.request_summary;
  if (envelope.repeated_count !== undefined) next.repeated_count = envelope.repeated_count;
  if (envelope.result_ref !== undefined && envelope.result_ref !== null) {
    next.result_ref = envelope.result_ref;
  }
  if (envelope.ended_at !== undefined && envelope.ended_at !== null) {
    next.ended_at = envelope.ended_at;
  }
  if (envelope.problem !== undefined) {
    next.problem = envelope.problem;
  }
  // Stamp ended_at on a terminal transition when the wire didn't carry one
  // (lean events omit it) so the Activity panel can compute elapsed time.
  if (isTerminalStatus(next.status) && next.ended_at === undefined) {
    next.ended_at = next.updated_at;
  }
  return next;
}

export const JOBS_STORAGE_KEY = 'andes-app:activity-v1';

export const useJobsStore = create<JobsState>()(
  persist(
    (set, get) => ({
      jobs: {},
      dismissedJobIds: [],

      addJob: (input) => {
        const id = input.id ?? mintLocalJobId();
        const now = Date.now() / 1000;
        const record: JobRecord = {
          id,
          kind: input.kind,
          status: input.status ?? 'pending',
          started_at: now,
          updated_at: now,
          can_cancel: input.can_cancel ?? false,
          request_summary: input.request_summary ?? {},
          repeated_count: 0,
          isPlaceholder: input.isPlaceholder ?? id.startsWith(LOCAL_ID_PREFIX),
        };
        if (input.progress !== undefined) record.progress = input.progress;
        const inserted = { ...get().jobs, [id]: record };
        set({ jobs: applyRetention(inserted) });
        return id;
      },

      updateJob: (id, patch) => {
        const cur = get().jobs[id];
        if (!cur) return;
        const next: JobRecord = {
          ...cur,
          ...patch,
          id: cur.id,
          updated_at: patch.updated_at ?? Date.now() / 1000,
        };
        if (
          patch.status !== undefined &&
          isTerminalStatus(patch.status) &&
          next.ended_at === undefined
        ) {
          next.ended_at = next.updated_at;
        }
        set({ jobs: applyRetention({ ...get().jobs, [id]: next }) });
      },

      upsertJob: (envelope) => {
        const existing = get().jobs[envelope.job_id];
        const merged = mergeEnvelope(envelope, existing);
        set({ jobs: applyRetention({ ...get().jobs, [envelope.job_id]: merged }) });
      },

      reconcileJob: (tempId, serverId, patch) => {
        const jobs = get().jobs;
        const placeholder = jobs[tempId];
        const serverRecord = jobs[serverId];
        const next = { ...jobs };

        if (serverRecord) {
          // The WS event raced ahead of the mutation response. Keep the
          // canonical server record; fold in any placeholder request_summary
          // we didn't already have, plus the response patch. Drop the temp.
          const merged: JobRecord = {
            ...serverRecord,
            ...(placeholder && Object.keys(serverRecord.request_summary).length === 0
              ? { request_summary: placeholder.request_summary }
              : {}),
            ...patch,
            id: serverId,
            updated_at: patch?.updated_at ?? Date.now() / 1000,
          };
          delete merged.isPlaceholder;
          next[serverId] = merged;
          if (tempId !== serverId) delete next[tempId];
        } else if (placeholder) {
          // Normal path: re-key the placeholder to the canonical id.
          const merged: JobRecord = {
            ...placeholder,
            ...patch,
            id: serverId,
            updated_at: patch?.updated_at ?? Date.now() / 1000,
          };
          delete merged.isPlaceholder;
          if (tempId !== serverId) delete next[tempId];
          next[serverId] = merged;
        } else {
          // Neither exists (defensive): synthesize a minimal record from the
          // patch so the response isn't dropped.
          const now = Date.now() / 1000;
          next[serverId] = {
            kind: (patch?.kind as JobKind) ?? 'pflow',
            status: patch?.status ?? 'done',
            started_at: patch?.started_at ?? now,
            updated_at: now,
            can_cancel: patch?.can_cancel ?? false,
            request_summary: patch?.request_summary ?? {},
            repeated_count: patch?.repeated_count ?? 0,
            ...patch,
            id: serverId,
          };
        }
        set({ jobs: applyRetention(next) });
      },

      removeJob: (id) => {
        if (!get().jobs[id]) return;
        const next = { ...get().jobs };
        delete next[id];
        set({ jobs: next });
      },

      syncJobs: (envelopes) => {
        // Full re-sync (reconnect / snapshot): the substrate payload is
        // authoritative for the jobs it lists. Preserve any still-IN-FLIGHT
        // record NOT present in the incoming set — that covers both raw
        // ``local:`` placeholders (server id not yet landed) AND optimistic
        // records the mutation path already reconciled onto a canonical id
        // but whose snapshot the server captured before registering. Dropping
        // them here would lose the optimistic request_summary an in-flight
        // row needs. Terminal records absent from the snapshot are genuinely
        // gone (evicted server-side) and are not preserved.
        const prev = get().jobs;
        const incoming = new Set(envelopes.map((e) => e.job_id));
        const next: Record<string, JobRecord> = {};
        for (const [id, rec] of Object.entries(prev)) {
          if (!incoming.has(id) && !isTerminalStatus(rec.status)) {
            next[id] = rec;
          }
        }
        for (const env of envelopes) {
          next[env.job_id] = mergeEnvelope(env, prev[env.job_id]);
        }
        set({ jobs: applyRetention(next) });
      },

      dismissJob: (id) => {
        if (get().dismissedJobIds.includes(id)) return;
        set({ dismissedJobIds: [...get().dismissedJobIds, id] });
      },

      undismissJob: (id) => {
        if (!get().dismissedJobIds.includes(id)) return;
        set({ dismissedJobIds: get().dismissedJobIds.filter((x) => x !== id) });
      },

      clearJobs: () => set({ jobs: {} }),
    }),
    {
      name: JOBS_STORAGE_KEY,
      storage: createJSONStorage(() => localStorage),
      // SECURITY (F2 / KTD-16): persist ONLY display-state metadata. The
      // full ``jobs`` map (request_summary payloads, problem error details,
      // result refs) is in-memory only and MUST NOT be serialized. This
      // whitelist is the single enforcement point — never widen it to
      // include ``jobs``.
      partialize: (state) => ({ dismissedJobIds: state.dismissedJobIds }),
    },
  ),
);

// Test-only: re-export internal helpers for retention / merge assertions
// without widening the public store API.
export const __internal = {
  applyRetention,
  mergeEnvelope,
  MAX_TERMINAL,
};
